File: C++/RasBot.cpp
#include "RasBot.h"

RasBot::RasBot(){}
RasBot::RasBot(int _mLF_port, int _mLB_port, int _mRF_port, int _mRB_port){
    m_LF = _mLF_port;
    m_LB = _mLB_port;
    m_RF = _mRF_port;
    m_RB = _mRB_port;

    #ifndef CHECK_CODE_OUTSIDE_RASP
    if(wiringPiSetup() == -1){
        #ifdef PRINT_DEBUG
        std::cout << "Erro config GPIO" << std::endl;
        #endif
        exit(1);
    }
    pinMode(m_LF, PWM_OUTPUT);
    pinMode(m_LB, PWM_OUTPUT);
    pinMode(m_RF, PWM_OUTPUT);
    pinMode(m_RB, PWM_OUTPUT);
    #endif
}
RasBot::~RasBot(){
}

bool RasBot::setPorts(int _mLF_port, int _mLB_port, int _mRF_port, int _mRB_port){
    m_LF = _mLF_port;
    m_LB = _mLB_port;
    m_RF = _mRF_port;
    m_RB = _mRB_port;

    #ifndef CHECK_CODE_OUTSIDE_RASP
    if(wiringPiSetup() == -1){
        #ifdef PRINT_DEBUG
        std::cout << "Config GPIO ... Error" << std::endl;
        #endif
        return false;
    }
    pinMode(m_LF, PWM_OUTPUT);
    pinMode(m_LB, PWM_OUTPUT);
    pinMode(m_RF, PWM_OUTPUT);
    pinMode(m_RB, PWM_OUTPUT);
    #endif

    #ifdef PRINT_DEBUG
    std::cout << "Config GPIO ... Done" << std::endl;
    #endif

    return true;
}

void RasBot::moveF(int relative_velocity){
    #ifdef PRINT_DEBUG
    std::cout << "Forward  -> ";
    #endif

    motorL(1, relative_velocity);
    motorR(1, relative_velocity);

    #ifdef PRINT_DEBUG
    std::cout << std::endl;
    #endif
}
void RasBot::moveB(int relative_velocity){
    #ifdef PRINT_DEBUG
    std::cout << "Backward -> ";
    #endif

    motorL(0, relative_velocity);
    motorR(0, relative_velocity);

    #ifdef PRINT_DEBUG
    std::cout << std::endl;
    #endif
}
void RasBot::turnL(int relative_velocity){
    #ifdef PRINT_DEBUG
    std::cout << "Left     -> ";
    #endif

    motorL(0, relative_velocity/2);
    motorR(1, relative_velocity/2);
    
    #ifdef PRINT_DEBUG
    std::cout << std::endl;
    #endif
}
void RasBot::turnR(int relative_velocity){
    #ifdef PRINT_DEBUG
    std::cout << "Right    -> ";
    #endif

    motorL(1, relative_velocity/2);
    motorR(0, relative_velocity/2);
    
    #ifdef PRINT_DEBUG
    std::cout << std::endl;
    #endif
}
void RasBot::motorL(int direction, int relative_velocity){
    if(direction){
        #ifndef CHECK_CODE_OUTSIDE_RASP
        digitalWrite(m_LB, LOW);
        pwmWrite(m_LF,relative_velocity);
        #endif
        #ifdef PRINT_DEBUG
        std::cout << " \t L:  " << relative_velocity;
        #endif
    }else{
        #ifndef CHECK_CODE_OUTSIDE_RASP
        digitalWrite(m_LF, LOW);
        pwmWrite(m_LB,relative_velocity);
        #endif
        #ifdef PRINT_DEBUG
        std::cout << " \t L: -" << relative_velocity;
        #endif
    }
}
void RasBot::motorR(int direction, int relative_velocity){
    if(direction){
        #ifndef CHECK_CODE_OUTSIDE_RASP
        digitalWrite(m_RB, LOW);
        pwmWrite(m_RF,relative_velocity);
        #endif
        #ifdef PRINT_DEBUG
        std::cout << " \t R:  " << relative_velocity;
        #endif
    }else{
        #ifndef CHECK_CODE_OUTSIDE_RASP
        digitalWrite(m_RF, LOW);
        pwmWrite(m_RB,relative_velocity);
        #endif
        #ifdef PRINT_DEBUG
        std::cout << " \t R: -" << relative_velocity;
        #endif
    }
}

File: Python/RasBot.py
PRINT_DEBUG = 1

import RPi.GPIO as gpio
from time import sleep

class RasBot:
    def __init__(self, mLF_port = 11, mLB_port = 12, mRF_port = 13, mRB_port = 15):
        gpio.cleanup()
        gpio.setmode(GPIO.BOARD)

        self.m_LF_port = mLF_port
        self.m_LB_port = mLB_port
        self.m_RF_port = mRF_port
        self.m_RB_port = mRB_port
        
        gpio.setup(self.m_LF_port, gpio.OUT)
        gpio.setup(self.m_LB_port, gpio.OUT)
        gpio.setup(self.m_RF_port, gpio.OUT)
        gpio.setup(self.m_RB_port, gpio.OUT)

        self.mLF = gpio.PWM(self.m_LF_port, 100)
        self.mLB = gpio.PWM(self.m_LB_port, 100)
        self.mRF = gpio.PWM(self.m_RF_port, 100)
        self.mRB = gpio.PWM(self.m_RB_port, 100)

        self.mLF.start(0)
        self.mLB.start(0)
        self.mRF.start(0)
        self.mRB.start(0)
    def setPorts(self, mLF_port, mLB_port, mRF_port, mRB_port):
        gpio.cleanup()
        gpio.setmode(GPIO.BOARD)

        self.m_LF_port = mLF_port
        self.m_LB_port = mLB_port
        self.m_RF_port = mRF_port
        self.m_RB_port = mRB_port
        
        gpio.setup(self.m_LF_port, gpio.OUT)
        gpio.setup(self.m_LB_port, gpio.OUT)
        gpio.setup(self.m_RF_port, gpio.OUT)
        gpio.setup(self.m_RB_port, gpio.OUT)

        self.mLF = gpio.PWM(self.m_LF_port, 100)
        self.mLB = gpio.PWM(self.m_LB_port, 100)
        self.mRF = gpio.PWM(self.m_RF_port, 100)
        self.mRB = gpio.PWM(self.m_RB_port, 100)

        self.mLF.start(0)
        self.mLB.start(0)
        self.mRF.start(0)
        self.mRB.start(0)
    def stop(self):
        self.mLF.ChangeDutyCycle(0)
        self.mLB.ChangeDutyCycle(0)
        self.mRF.ChangeDutyCycle(0)
        self.mRB.ChangeDutyCycle(0)
        sleep(0.02)
    def motorL(self, direction, relative_velocity):
        if direction:
            self.mLB.ChangeDutyCycle(0)
            self.mLF.ChangeDutyCycle(relative_velocity)
            if PRINT_DEBUG:
                print(' R:  {}'.format(relative_velocity))
        else:
            self.mLF.ChangeDutyCycle(0)
            self.mLB.ChangeDutyCycle(relative_velocity)
            if PRINT_DEBUG:
                print(' R: -{}'.format(relative_velocity))
    def motorR(self, direction, relative_velocity):
        if direction:
            self.mRB.ChangeDutyCycle(0)
            self.mRF.ChangeDutyCycle(relative_velocity)
            if PRINT_DEBUG:
                print(' R:  {}'.format(relative_velocity))
        else:
            self.mRF.ChangeDutyCycle(0)
            self.mRB.ChangeDutyCycle(relative_velocity)
            if PRINT_DEBUG:
                print(' R: -{}'.format(relative_velocity))
    def moveF(self, relative_velocity):
        self.motorL(1,relative_velocity)
        self.motorR(1,relative_velocity)
        if PRINT_DEBUG:
                print('-Forward-')
    def moveB(self, relative_velocity):
        self.motorL(0,relative_velocity)
        self.motorR(0,relative_velocity)
        if PRINT_DEBUG:
                print('-Backward-')
    def turnL(self, relative_velocity):
        self.motorL(0,relative_velocity)
        self.motorR(1,relative_velocity)
        if PRINT_DEBUG:
                print('-Left-')
    def turnR(self, relative_velocity):
        self.motorL(1,relative_velocity)
        self.motorR(0,relative_velocity)
        if PRINT_DEBUG:
                print('-Right-')